File: examples/measurements.c
/* SPDX-License-Identifier: BSD-3-Clause-Clear */
/* Application composition example; not a new installed API or USB owner.
 * Caller freshly boots pinned firmware, tunes 2.4GHz channel6/20MHz, then starts
 * one mt_session. No concurrent retunes/consumers/callers during this function.
 * After return, caller must stop then destroy session BEFORE closing the device.
 * A failed stop retains ownership: do not free/close; retry orderly stop.
 * Build check: clang -Wall -Wextra -Werror -std=c11 -Ic -c examples/measurements.c
 * Link with the same driver objects/frameworks as c/mt76_session_probe.
 */
#include "mt76_session.h"
#include "mt7921_radio.h"
#include <inttypes.h>

typedef struct {
    mt_counter_sample_t counters;
    mt_thermal_sample_t thermal;
} example_sample_t;

static int measure(mt7921_dev_t *dev, void *ctx) {
    example_sample_t *sample = ctx;
    const int names[] = {MT_COUNTER_RX_MPDU, MT_COUNTER_PRIMARY_CCA};
    if (mt_counter_read(dev, names, 2, &sample->counters)) return -1;
    return mt_thermal_read(&dev->mcu, MT_THERMAL_TEMPERATURE, &sample->thermal);
}

/* 0 completed with RX, 2 completed without RX (inconclusive), -1 error.
 * Output is aggregate NDJSON, never packet bytes or ambient identifiers.
 * Classified RX packets are not a good-FCS decode guarantee.
 */
int example_measurements(mt76_session_t *session, unsigned seconds, FILE *output) {
    if (!session || !output || seconds < 1 || seconds > 60) return -1;
    uint64_t started = mt_radio_monotonic_us(), next = started, frames = 0;
    mt_session_packet_t packet;
    while (mt_radio_monotonic_us() - started < (uint64_t)seconds * 1000000) {
        int result = mt_session_read(session, &packet, false, 50);
        if (result < 0) return -1;
        if (!result) frames++;
        while (!mt_session_read(session, &packet, true, 0)) {}
        mt_session_stats_t before, after;
        mt_session_snapshot(session, &before);
        if (before.state != MT_SESSION_RUNNING) return -1;
        if (mt_radio_monotonic_us() < next) continue;
        example_sample_t sample;
        if (mt_session_call(session, measure, &sample, 3000, false)) return -1;
        mt_session_snapshot(session, &after);
        if (after.state != MT_SESSION_RUNNING || after.epoch_ns != before.epoch_ns ||
            after.generation != before.generation) return -1;
        /* RX_MPDU is a count; PRIMARY_CCA has unknown duration scaling.
         * Counter interval and thermal interval are separate host observations.
         */
        if (fprintf(output,
            "{\"event\":\"measurement\",\"schema\":1,\"epoch_ns\":%" PRIu64
            ",\"channel_generation\":%u,\"requested_control\":%u,"
            "\"frames_consumed\":%" PRIu64 ",\"counter_opened_us\":%" PRIu64
            ",\"counter_closed_us\":%" PRIu64 ",\"counters\":["
            "{\"name\":\"%s\",\"raw\":%" PRIu64 ",\"unit\":%d,\"tick_ns\":null},"
            "{\"name\":\"%s\",\"raw\":%" PRIu64 ",\"unit\":%d,\"tick_ns\":null}],"
            "\"reported_temperature_c\":%" PRId32 ",\"thermal_opened_us\":%" PRIu64
            ",\"thermal_closed_us\":%" PRIu64 ",\"channel_busy_fraction\":null,"
            "\"frames_dropped\":%" PRIu64 ",\"events_dropped\":%" PRIu64
            ",\"usb_errors\":%" PRIu64 "}\n",
            after.epoch_ns, after.generation, after.control, frames,
            sample.counters.raw.opened_us, sample.counters.raw.closed_us,
            sample.counters.descriptors[0]->name, sample.counters.raw.values[0],
            sample.counters.descriptors[0]->unit,
            sample.counters.descriptors[1]->name, sample.counters.raw.values[1],
            sample.counters.descriptors[1]->unit,
            sample.thermal.reported_temperature_c, sample.thermal.opened_us,
            sample.thermal.closed_us, after.frames_dropped, after.events_dropped,
            after.usb_errors) < 0 || fflush(output)) return -1;
        next = mt_radio_monotonic_us() + 1000000;
    }
    return frames ? 0 : 2;
}

File: examples/measurements.py
#!/usr/bin/env python3
# SPDX-License-Identifier: BSD-3-Clause-Clear
"""Compose installed APIs: passive channel6 capture and raw interval measurements.

Run after installing the package: python examples/measurements.py --usb-id 0846:9072
Uses checksum-pinned firmware from MT76_FW_DIR. No research imports or RF transmit.
Output contains aggregate counts only, not frames, addresses, SSIDs or CSI arrays.
"""

from __future__ import annotations

import argparse
import hashlib
import json
import time

import mt7921u as m
from mt76_measurements import Counter, read_counters, read_thermal
from mt76_session import AcquisitionSession, SessionError


def collect(session, seconds=10, emit=print):
    """Consume one caller-owned running session; caller must stop it before USB close.

    No concurrent retunes, consumers or measurement callers in this small example.
    A callback serializes USB commands, not the hardware sampling clocks. The C
    equivalent is examples/measurements.c. Errors propagate; never export a fake zero.
    """
    if type(seconds) is not int or not 1 <= seconds <= 60:
        raise ValueError("seconds must be an integer in 1..60")
    started = time.monotonic()
    next_sample = started
    frames = 0
    while time.monotonic() - started < seconds:
        packet = session.read(timeout=0.05)
        if packet is not None:
            frames += 1  # Classified RX packets, not a good-FCS decode guarantee.
        while session.read(timeout=0, events=True) is not None:
            pass
        before = session.snapshot()
        if before["state"] != "running":
            raise SessionError("acquisition worker stopped")
        if time.monotonic() < next_sample:
            continue
        counters, thermal = session.call(
            lambda dev: (
                read_counters(dev, (Counter.RX_MPDU, Counter.PRIMARY_CCA)),
                read_thermal(dev),
            )
        )
        after = session.snapshot()
        if (
            after["state"] != "running"
            or after["epoch_ns"] != before["epoch_ns"]
            or after["channel_generation"] != before["channel_generation"]
        ):
            raise SessionError("measurement context changed")
        emit(
            json.dumps(
                {
                    "event": "measurement",
                    "schema": 1,
                    "epoch_ns": after["epoch_ns"],
                    "channel_generation": after["channel_generation"],
                    "requested_channel": after["requested_channel"],
                    "frames_consumed": frames,
                    "counter_opened_us": counters.opened_us,
                    "counter_closed_us": counters.closed_us,
                    "counters": [
                        {
                            "name": r.descriptor.name,
                            "raw": r.raw,
                            "unit": int(r.descriptor.unit),
                            "tick_ns": r.descriptor.tick_ns,
                        }
                        for r in counters.readings
                    ],
                    "reported_temperature_c": thermal.reported_temperature_c,
                    "thermal_opened_us": thermal.opened_us,
                    "thermal_closed_us": thermal.closed_us,
                    "channel_busy_fraction": None,  # Unknown duration scaling/coverage.
                    "session": after,  # Queue loss and command failures remain visible.
                }
            )
        )
        next_sample = time.monotonic() + 1
    return frames


def main(argv=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--usb-id", required=True, choices=("0e8d:7961", "0846:9072"))
    parser.add_argument("--seconds", type=int, choices=range(1, 61), default=10)
    args = parser.parse_args(argv)
    with m.open_device(args.usb_id) as dev:
        patch, ram = m.load_firmware(dev.CHIP)
        dev.bringup(patch, ram, log=lambda *_: None)
        dev.set_monitor_mode()
        dev.set_sniffer(True)
        dev.tune("2.4GHz", 6)
        print(
            json.dumps(
                {
                    "event": "ready",
                    "chip": dev.CHIP,
                    "patch_sha256": hashlib.sha256(patch).hexdigest(),
                    "ram_sha256": hashlib.sha256(ram).hexdigest(),
                }
            ),
            flush=True,
        )
        # The inner context joins the USB worker before the outer closes USB.
        with AcquisitionSession(dev) as session:
            frames = collect(session, args.seconds)
    return 0 if frames else 2  # Zero received packets is inconclusive, not an RX pass.


if __name__ == "__main__":
    raise SystemExit(main())
